R/mixMunsell.R

```diff
@@ -149,52 +180,54 @@ mixMunsell <- function(x, w = rep(1, times = length(x)) / length(x), n = 1) {
     )
     return(res)
   }
-  
+
   # empty vector for mixture
   mixed <- vector(mode = 'numeric', length = nrow(s))
-  
+
   # iterate over wavelength (columns in first spectra)
   for(i in seq_along(mixed)) {
-    
+
     # prepare values:
     # select the i-th wavelength (row)
     # down-grade to a vector
     vals <- unlist(s[i, ])
-    
+
+    # aggregate weights by "chip" -- in case length(x) != length(unique(x))
+    wagg <- aggregate(w, by = list(chip = x), FUN = sum)
+
     # mix via weighted geometric mean
-    mixed[i] <- .wgm( v = vals, w = w )
+    mixed[i] <- .wgm(v = vals, w = wagg$x[match(names(s), wagg$chip)])
   }
-  
-  
+
   ## TODO: what is the "best" distance metric when comparing 1 spectra to the entire library?
-  
+
   ## optimization: matrixStats::colSums2() much faster
   ## --> syntax slightly different
-  
+
   ## for now, using Euclidean distance
   # D = sqrt(sum( [reference - mixed]^2 ))
-  
+
   # subtract the mixture spectra, element-wise, from reference library
   # note we are removing the wavelength (1st) column
   m.diff <- sweep(munsell.spectra.wide[, -1], MARGIN = 1, STATS = mixed, FUN = '-')
-  
+
   # finish the distance calculation
   m.dist <- sqrt(colSums(m.diff^2))
-  
+
   # get the spectra of the closest n munsell chip(s) via sorting ASC
   m.match <- sort(m.dist)[1:n]
-  
+
   # compile into data.frame
   res <- data.frame(
     munsell = names(m.match),
     distance = m.match,
     stringsAsFactors = FALSE
   )
-  
+
   # clean-up row names
-  row.names(res) <- as.character(1:nrow(res))
-  
+  row.names(res) <- NULL
```

Comment:
Hey that is cool, learned something new.

Reply:
It is cool, save a couple key strokes. 

Not that it matters for this function but non-`NULL` `row.names<-` to a [`tibble` causes warnings]( https://tibble.tidyverse.org/reference/rownames.html). I think the thought is that they get do get messy, as above, and if they contain any real info they should be a column.

Reply:
Good point. I'll be sure to use this construct moving forward.